examples/streaming_chatbot/application.py

```diff
@@ -0,0 +1,187 @@
+from typing import List, Optional, Tuple
+
+import openai
+
+from burr.core import ApplicationBuilder, State, default, when
+from burr.core.action import action, streaming_action
+from burr.lifecycle import LifecycleAdapter
+
+MODES = {
+    "answer_question": "text",
+    "generate_image": "image",
+    "generate_code": "code",
+    "unknown": "text",
+}
+
+
+@action(reads=[], writes=["chat_history", "prompt"])
+def process_prompt(state: State, prompt: str) -> Tuple[dict, State]:
+    result = {"chat_item": {"role": "user", "content": prompt, "type": "text"}}
+    return result, state.wipe(keep=["prompt", "chat_history"]).append(
+        chat_history=result["chat_item"]
+    ).update(prompt=prompt)
+
+
+@action(reads=["prompt"], writes=["safe"])
+def check_safety(state: State) -> Tuple[dict, State]:
+    result = {"safe": "unsafe" not in state["prompt"]}  # quick hack to demonstrate
+    return result, state.update(safe=result["safe"])
+
+
+def _get_openai_client():
+    return openai.Client()
+
+
+@action(reads=["prompt"], writes=["mode"])
+def choose_mode(state: State) -> Tuple[dict, State]:
+    prompt = (
+        f"You are a chatbot. You've been prompted this: {state['prompt']}. "
+        f"You have the capability of responding in the following modes: {', '.join(MODES)}. "
+        "Please respond with *only* a single word representing the mode that most accurately "
+        "corresponds to the prompt. Fr instance, if the prompt is 'draw a picture of a cat', "
+        "the mode would be 'generate_image'. If the prompt is 'what is the capital of France', the mode would be 'answer_question'."
+        "If none of these modes apply, please respond with 'unknown'."
+    )
+
+    result = _get_openai_client().chat.completions.create(
+        model="gpt-4",
+        messages=[
+            {"role": "system", "content": "You are a helpful assistant"},
+            {"role": "user", "content": prompt},
+        ],
+    )
+    content = result.choices[0].message.content
+    mode = content.lower()
+    if mode not in MODES:
+        mode = "unknown"
+    result = {"mode": mode}
+    return result, state.update(**result)
+
+
+@action(reads=["prompt", "chat_history"], writes=["response"])
+def prompt_for_more(state: State) -> Tuple[dict, State]:
+    """Not streaming, as we have the result immediately."""
+    result = {
+        "response": {
+            "content": "None of the response modes I support apply to your question. Please clarify?",
+            "type": "text",
+            "role": "assistant",
+        }
+    }
+    return result, state.update(**result).append(chat_history=result["response"])
+
+
+@streaming_action(reads=["prompt", "chat_history", "mode"], writes=["response"])
+def chat_response(
+    state: State, prepend_prompt: str, model: str = "gpt-3.5-turbo"
+) -> Tuple[dict, State]:
```

Comment:
this has the wrong annotation -- shouldn't we error here?

Reply:
Not compiling against annotations now -- its just a type-hint for the user (and the IDE should be smart enough to know)

Reply:
Fixed the annotation